Workflow/Sources/WorkflowObserver.swift

```diff
@@ -0,0 +1,322 @@
+/*
+ * Copyright 2022 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+
+// MARK: - WorkflowObserver
+
+/// The `WorkflowObserver` protocol provides an interface to receive updates during the runtime's
+/// execution loop. All requirements are optional, and no-ops are provided by default.
```

Comment:
Xcode 14 appears to provide a somewhat-improved check for simple typos – it will warn if an instance method 'nearly matches' a defaulted requirement. some fairly straightforward unit tests of any conformance would likely suffice as defense against the typo concern, i think.

i'm not convinced that if you adopt this protocol you'd inherently want to observe everything. for example, if you wanted to implement separate observers to handle different responsibilities (e.g. one for measuring render times, one for recording action emissions, etc), then you'd have a fair amount of boilerplate to provide no-op implementations of the irrelevant bits. i'll think on it a bit more though...

Reply:
OK that's fair 👍 (sorry I totally missed your responses before, I need to figure out my email filters…)